File: fundamentals/p412/problem-412.cpp
/**
 * @file problem-412.cpp
 * @brief LeetCode Problem 412: Fizz Buzz (Standard Approach)
 * 
 * @note
 * - Documentation: Created using LLM / AI.
 * - Solution Origin: Human written logic.
 * 
 * @details
 * Given an integer `n`, return a string array answer (1-indexed) where:
 * - answer[i] == "FizzBuzz" if i is divisible by 3 and 5.
 * - answer[i] == "Fizz" if i is divisible by 3.
 * - answer[i] == "Buzz" if i is divisible by 5.
 * - answer[i] == i (as a string) if none of the above conditions are true.
 * 
 * Algorithm: Modulo Condition Checks
 * - Loop from 1 to `n`.
 * - Perform standard checks using the modulo operator `%` for divisibility by 3 and 5.
 * - Push corresponding strings to the answer vector.
 * 
 * Time Complexity:  O(N) - Loop runs exactly N times.
 * Space Complexity: O(1) auxiliary - No extra memory besides the output array.
 */

#include <iostream>
#include <vector>
#include <string>
using namespace std;

class Solution {
public:
    /**
     * @brief Generates FizzBuzz sequence from 1 to n.
     * @param n Maximum limit.
     * @return vector<string> List of FizzBuzz results.
     */
    vector<string> fizzBuzz(int n) {
        vector<string> answer;
        for(int i = 1; i <= n; i++) {
            if(i % 3 == 0 && i % 5 == 0)
                answer.push_back("FizzBuzz");
            else if(i % 3 == 0)
                answer.push_back("Fizz");
            else if(i % 5 == 0)
                answer.push_back("Buzz");
            else
                answer.push_back(to_string(i));
        }
        return answer;
    }
};

int main() {
    Solution s;
    int n = 15;
    vector<string> result = s.fizzBuzz(n);
    
    cout << "FizzBuzz (" << n << "):" << endl;
    for (const string& str : result) {
        cout << str << " ";
    }
    cout << endl;
    
    return 0;
}

File: 1-arrays/p41/Problem41-2.java
/**
 * @file Problem41-2.java
 * @brief LeetCode Problem 41: First Missing Positive (Boolean Frequency/Presence Array Approach)
 * 
 * @note
 * - Documentation: Created using LLM / AI.
 * - Solution Origin: Human written logic.
 * 
 * @details
 * Given an unsorted integer array `nums`, return the smallest positive integer that is not present in `nums`.
 * 
 * Algorithm: Boolean Presence Array
 * - The first missing positive must lie in the range `[1, nums.length + 1]`.
 * - Allocate a boolean array `present` of size `nums.length + 1`.
 * - Mark `present[x] = true` for every `x` where `0 < x < present.length`.
 * - Scan from index `1` upwards; the first index with `!present[i]` is the answer.
 * - If all indices `1` through `nums.length` are present, return `present.length` (`nums.length + 1`).
 * 
 * Time Complexity:  O(N) - Two sequential linear passes.
 * Space Complexity: O(N) - Auxiliary space for the boolean presence array.
 */

public class Problem41_2 {
    /**
     * Finds the first missing positive integer using a boolean presence array.
     * @param nums Unsorted input integer array.
     * @return Smallest missing positive integer.
     */
    public int firstMissingPositive(int[] nums) {
        boolean[] present = new boolean[nums.length + 1];
        for (int x : nums) {
            if (x > 0 && x < present.length) {
                present[x] = true;
            }
        }
        for (int i = 1; i < present.length; i++) {
            if (!present[i])
                return i;
        }
        return present.length;
    }

    public static void main(String[] args) {
        Problem41_2 sol = new Problem41_2();
        int[] nums = {1, 2, 0};
        System.out.println("First missing positive: " + sol.firstMissingPositive(nums));
    }
}
